src/Compilers/VisualBasic/Portable/BoundTree/Statement.vb

```diff
@@ -636,29 +636,61 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
                             ' Use the operator methods. Figure out the precise rules first.
                             Return Nothing
                         Else
-                            If BoundFor.StepValue Is Nothing OrElse (BoundFor.StepValue.IsConstant AndAlso BoundFor.StepValue.ConstantValueOpt IsNot Nothing) Then
+                            ' Note: this is the path where were not using any user defined operators.
+                            ' i.e. all operators will be built in intrinsics, or lifted versions thereof.
+
+                            If BoundFor.StepValue Is Nothing OrElse
+                               (BoundFor.StepValue.IsConstant AndAlso BoundFor.StepValue.ConstantValueOpt IsNot Nothing) Then
                                 ' Either ControlVariable <= LimitValue or ControlVariable >= LimitValue, depending on whether the step value is negative.
 
-                                Dim relationalCode As BinaryOperationKind = DeriveBinaryOperationKind(If(BoundFor.StepValue IsNot Nothing AndAlso BoundFor.StepValue.ConstantValueOpt.IsNegativeNumeric, BinaryOperatorKind.GreaterThanOrEqual, BinaryOperatorKind.LessThanOrEqual), controlVariable)
-                                Return New Binary(relationalCode, controlVariable, limitValue, booleanType, limitValue.Syntax)
+                                Dim relationalCode As BinaryOperationKind = DeriveBinaryOperationKind(
+                                    If(BoundFor.StepValue IsNot Nothing AndAlso BoundFor.StepValue.ConstantValueOpt.IsNegativeNumeric,
+                                       BinaryOperatorKind.GreaterThanOrEqual,
+                                       BinaryOperatorKind.LessThanOrEqual),
+                                    controlVariable)
+
+                                ' We are comparing the control variable against the limit value.  Using
+                                ' either the default stepping constant, or a user supplied constant.
+                                ' This will be a lifted comparison if either the control variable or
+                                ' limit value is nullable itself.
+                                Dim isLifted = controlVariable.Type.OriginalDefinition.SpecialType = SpecialType.System_Nullable_T OrElse
+                                               limitValue.Type.OriginalDefinition.SpecialType = SpecialType.System_Nullable_T
```

Comment:
Here and below, please use IsNullableType helper instead of explicitly digging through OriginalDefinition